sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/models/HttpResponse.java

```diff
@@ -3,90 +3,145 @@
 
 package com.generic.core.http.models;
 
+import com.generic.core.http.Response;
 import com.generic.core.models.BinaryData;
-import com.generic.core.models.HeaderName;
 import com.generic.core.models.Headers;
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.util.function.Function;
 
 /**
  * The response of an {@link HttpRequest}.
  */
-public abstract class HttpResponse implements Closeable {
+public class HttpResponse<T> implements Response<T>, Closeable {
+    protected static final BinaryData EMPTY_BODY = BinaryData.fromBytes(new byte[0]);
+
+    private final Function<Response<?>, ?> deserializationCallback;
+    private final Headers headers;
     private final HttpRequest request;
-    private BinaryData binaryData = null;
-    private final byte[] bodyBytes;
+    private final int statusCode;
+
+    private BinaryData body;
+    private T value;
 
     /**
      * Creates an instance of {@link HttpResponse}.
      *
      * @param request The {@link HttpRequest} that resulted in this {@link HttpResponse}.
+     * @param statusCode The response status code.
+     * @param headers The response {@link Headers}.
+     * @param value The response body.
      */
-    protected HttpResponse(HttpRequest request) {
+    public HttpResponse(HttpRequest request, int statusCode, Headers headers, T value) {
         this.request = request;
-        this.bodyBytes = null;
+        this.statusCode = statusCode;
+        this.headers = headers;
+        this.value = value;
+        this.deserializationCallback = request == null
+            ? Response::getValue
+            : request.getResponseBodyDeserializationCallback();
     }
 
     /**
      * Creates an instance of {@link HttpResponse}.
      *
      * @param request The {@link HttpRequest} that resulted in this {@link HttpResponse}.
-     * @param bodyBytes The response body as a byte array.
+     * @param statusCode The response status code.
+     * @param headers The response {@link Headers}.
+     * @param value The response body.
+     * @param isRawResponse Indicates whether the value provided needs to be deserialized.
      */
-    protected HttpResponse(HttpRequest request, byte[] bodyBytes) {
+    @SuppressWarnings("unchecked")
+    public HttpResponse(HttpRequest request, int statusCode, Headers headers, BinaryData value, boolean isRawResponse) {
```

Comment:
Does this need to be public ?

Reply:
No, I'm moving it to the implementation package.

Reply:
Sorry, I thought you meant the class, not the constructor. We can remove this constructor.